### DISCUSSION TITLE
Prop Diffing on the `TileLayer`

### DISCUSSION THREAD

**Discussion by ilan-gold** [visgl/deck.gl#6518] [Q&A] [answered]
We are having some issues with prop diffing on the `TileLayer` and are wondering if anyone has noticed similarly bizarre behavior with regards to this.  We tried making something minimal to replicate the issue with our layers but were unable to do so, so I'm more just posting this to see if anyone else has seen this since 8.6.0 (@Pessimistress I think it might have to do with the PR we did).  I'll add an example once I have one but having a bit of trouble doing that at the moment.

For example, I had `onViewportLoad` callbacks that updated some props that are passsed down to the layer but I had to move the prop updates outside the `onViewportLoad` function in order for the layer to update https://github.com/hms-dbmi/viv/pull/555/files - otherwise the layer never saw the updates

I'll try to post something more reproducible as I find it but we're having trouble even doing that.

**Comment by ilan-gold** [comment]
For example:

```
git clone https://github.com/hms-dbmi/vizarr.git
cd vizarr
gh pr checkout 126
npm i && npm start
```
And then paste http://localhost:3000/?source=https://minio-dev.openmicroscopy.org/idr/v0.3/idr0062-blin-nuclearsegmentation/6001240.zarr into your browser.  When you change slider labeled "z" nothing happens when in reality, the image should update.  And once you interact with the image via zoom/pan, it does update.

<img width="1440" alt="Screen Shot 2021-12-17 at 7 04 12 PM" src="https://user-images.githubusercontent.com/43999641/146588586-ba06b555-a9aa-4052-93b3-e5b93c19c3a8.png">

Expected behavior: https://hms-dbmi.github.io/vizarr/?source=https://minio-dev.openmicroscopy.org/idr/v0.3/idr0062-blin-nuclearsegmentation/6001240.zarr

**Reply by ilan-gold** [reply]
For reference, @manzt tried writing up a `gist` but was unable to reproduce the issue:

https://gist.github.com/manzt/c07aed66d2df12fc025d9c90dbc2d94a

I would think it's the state management library he was using if it weren't for the fact that I ran into a similar issue with a different library that uses deck.gl (although not an identical issue).

**Reply by ilan-gold** [reply]
It's possible this is a bug on our end!  Just trying to figure that out!

**Reply by ilan-gold** [reply]
We can see the prop being passed down to the layer - but the layer doesn't update (it doesn't even call things like `shouldUpdateState`).

**Reply by Pessimistress** [reply]
Can you point me to the source of your custom layer? Maybe update is blocked by `shouldUpdateState`?

**Reply by ilan-gold** [reply]
Ahhh! @Pessimistress  you are a blessing! https://github.com/hms-dbmi/viv/blob/master/src/layers/MultiscaleImageLayer/MultiscaleImageLayer.js wraps https://github.com/hms-dbmi/viv/blob/master/src/layers/MultiscaleImageLayer/MultiscaleImageLayerBase.js where we have only updated `_updateTileset` - we don't override `shouldUpdateState` in either.

**Reply by ilan-gold** [reply]
So it seems that this PR https://github.com/hms-dbmi/viv/pull/516 is what broke things ins `viv` - I tried progressing forward on `master` and everything up until this PR works.  The repo I shared for reproduction uses `viv` (the repo that is in the links to the layers above) as a library for rendering.

I refactored a lot of functionality onto `extensions` in that and I wonder if this is related to https://github.com/visgl/deck.gl/discussions/6310

**Reply by ilan-gold** [reply]
So the other place you notice this issue:

```
git clone https://github.com/hms-dbmi/viv.git
cd viv
npm i && npm start
````

Then paste in http://localhost:3000/?image_url=https://viv-demo.storage.googleapis.com/2018-12-18_ASY_H2B_bud_05_3D_8_angles.ome.tif to your browser.  Change the z-stack forward to 238
<img width="1438" alt="Screen Shot 2021-12-17 at 8 24 15 PM" src="https://user-images.githubusercontent.com/43999641/146597415-24a44f54-b2a1-48db-bc68-3d9fc28c6ce9.png">

and then back in the other direction and screen goes white
<img width="1440" alt="Screen Shot 2021-12-17 at 8 24 25 PM" src="https://user-images.githubusercontent.com/43999641/146597467-c2157ad1-37bd-4c4d-ad63-8756106fdbba.png">


the reason for this is that when you jump to 238 in the z-stack selection, the image data is empty so the contrast settings are set to `[0,0]` automatically.  Then when you switch back, for some reason, deck.gl is one update behind so only applies those `[0,0]` settings to this final result which is what causes it look white even though they were the settings for the last image.  When I do a stack trace or `console.log` I can see the contrast settings being passed down, but deck.gl reacting one selection event behind the actual selection.

I know this is a lot but I think the gist is that this seems somehow related to prop diffing and extensions, which I mentioned before: https://github.com/visgl/deck.gl/discussions/6310

**Reply by Pessimistress** [reply]
It is a known issue that extension prop types are not utilized during diffing. But that issue will only lead to false positives (i.e. too many prop changes) not false negatives.

I don't see any non-standard props passed to the TileLayer in your source. Which prop is causing the problem?

**Reply by ilan-gold** [reply]
TL;DR I think this is somehow related to #6310 - sorry for the flood of messages.

**Reply by ilan-gold** [reply]
> It is a known issue that extension prop types are not utilized during diffing. But that issue will only lead to false positives (i.e. too many prop changes) not false negatives.
> 
> I don't see any non-standard props passed to the TileLayer in your source. Which prop is causing the 

I think all props are equally not being seen as updated.  I had to remove the automatic prop setting outside the `onViewportLoad` callback to fix this issue: https://github.com/hms-dbmi/viv/pull/555

It didn't make a difference which prop was set in the callback, they all never updated.

**Reply by ilan-gold** [reply]
Specficially any state that I set in `setPropertiesForChannel` in these lines was not seen by the layer: https://github.com/hms-dbmi/viv/blob/17154348bf71676ac73f182a5f07c1df927176b8/avivator/src/components/Controller/components/GlobalSelectionSlider.jsx#L33-L51

**Reply by ilan-gold** [reply]
This could be an issue with state management libraries but @manzt and I use different libraries (I use `zustand` and he `jotai`)

**Reply by Pessimistress** [reply]
I suspect it's an issue in the state management system. Can you print out the layers during each render and inspect their props?

**Reply by ilan-gold** [reply]
@Pessimistress If you could point me to a breakpoint to set in the deck.gl library that would be great - you of course don't have to dig into this.  For example, I was able to step into the construction of the layer and I could see the "new" prop there but then when I log things out in the `renderSubLayers` of the layer itself, I don't see it.

**Reply by ilan-gold** [reply]
> I suspect it's an issue in the state management system. Can you print out the layers during each render and inspect their props?

I did that exactly, yes! I simply never see the prop and then If I touch the canvas with zoom or pan, I do see the prop update.

**Reply by ilan-gold** [reply]
@Pessimistress Let me share a screen grab movie.

**Reply by Pessimistress** [reply]
Try https://github.com/visgl/deck.gl/blob/master/modules/core/src/lib/deck.js#L260 - the Layer instance you construct may not be the one you pass to DeckGL.

**Comment by Pessimistress** [comment, marked as answer]
Fix published in v8.6.7.
